etna/ciim/models.py

```diff
@@ -1,237 +1,77 @@
-from abc import ABC
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
 
 from django.conf import settings
+from django.utils.functional import classproperty
 
 from .client import KongClient
-from .exceptions import (
-    DoesNotExist,
-    InvalidQuery,
-    KongAPIError,
-    MultipleObjectsReturned,
-    UnsupportedSlice,
-)
-
-
-class SearchManager:
-    """A model.Manager/QuerySet-like object to query and fetch KongModel instances."""
-
-    def __init__(self, model):
-        self.model = model
-
-    def filter(self, **kwargs):
-        # Check we're making a meaningful query before contacting Kong
-        if any({k: v for k, v in kwargs.items() if v is None}):
-            raise InvalidQuery("Query to Kong must at least one clause")
-
-        return SearchQuery(self.model, kwargs)
-
-    def get(self, iaid=None, reference_number=None, expand=None):
-        return FetchQuery(self.model).get(iaid=iaid, reference_number=reference_number, expand=expand)
-
-    def get_multiple(self, iaid=None):
-        return FetchQuery(self.model).get_multiple(iaid=iaid)
+from .exceptions import DoesNotExist, MultipleObjectsReturned
 
 
-class ResultPending(Exception):
-    """Raised if result being requested that hasn't yet been fetched."""
+class APIModel(ABC):
+    """Model representation of API data.."""
 
+    @classmethod
+    @abstractmethod
+    def from_api_response(cls, response: dict) -> APIModel:
+        """Transform a response From Kong into an APIModel instance.
 
-class ResultSet:
-    """Wrapper for a list of ES search results.
-
-    Contains a list the size of the ES results set with each item being None or
-    a dict.
-
-    If None is accessed from ResultSet, client is informed that the result
-    hasn't yet been fetched from ES.
-    """
-
-    def __init__(self, total_count=0):
-        self.results = [None] * total_count
-
-    def __getitem__(self, index):
-        """Support subscripting.
-
-        Raise IndexError if requested item is not in list
-        or ResultPending if item hasn't been fetched yet.
+        To be implemented the concrete class.
         """
-        result = self.results[index]
-        if not result:
-            raise ResultPending
+        raise NotImplementedError
+
+    @classproperty
+    def search(cls):
+        raise DeprecationWarning(
+            f"{cls.__name__}.search is deprecated. Use {cls.__name__}.api instead."
+        )
 
-        return result
+    @classproperty
+    def api(cls) -> APIManager:
```

Comment:
I think that the model needs No access to the APIManager, for me, this is the separation of concern , and all api calls are within the concern of the APIManager, for example I don't think we should do something like that : 
page = Record.api.fetch(iaid=iaid, expand=True).

Reply:
This is a similar pattern to the way we'd use Django's ORM (`Model.objects`)

https://docs.djangoproject.com/en/4.0/topics/db/queries/#retrieving-objects

`Records.api` is just a convenient way of accessing a preconfigured `APIManager` instance, ready to fetch and transform `Record` data.

Reply:
Thank you Dan :) , 
But the actual question and form design point view, why the model is performing the Kong api search, It is slightly breaking MVC pattern, in MVC a model is only responsible for the data definition, required or not, length, relationship with the other models, primary key and foreign, ...,

Reply:
I understand your concern, this might worth a chat if you want to discuss further.

The Model isn't performing the fetch for data. 

```python
records = Record.api.fetch(iaid=iaid, expand=True)
```

Is shorthand for:

```python
manager = APIManager(Record)
records = manager.fetch(iaid=iaid, expand=True)
```

Reply:
Thank you Dan.
Maybe this is my last comment on this, I believe the Model must not reference the APIManger.